vectorstores/alloydb/vectorestores.go

```diff
@@ -0,0 +1,346 @@
+package alloydb
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v5"
+	"github.com/tmc/langchaingo/embeddings"
+	"github.com/tmc/langchaingo/internal/alloydbutil"
+	"github.com/tmc/langchaingo/schema"
+	"github.com/tmc/langchaingo/vectorstores"
+)
+
+const (
+	defaultIndexNameSuffix = "langchainvectorindex"
+	similaritySearchQuery  = `WITH filtered_embedding_dims AS MATERIALIZED (
+    SELECT
+        *
+    FROM
+        %s
+    WHERE
+        vector_dims (
+                embedding
+        ) = $1
+)
+SELECT
+    data.document,
+    data.cmetadata,
+    (1 - data.distance) AS score
+FROM (
+    SELECT
+        filtered_embedding_dims.*,
+        embedding <=> $2 AS distance
+    FROM
+        filtered_embedding_dims
+    JOIN %s ON filtered_embedding_dims.collection_id=%s.uuid
+    WHERE %s.name='%s') AS data
+WHERE %s
+ORDER BY
+    data.distance
+LIMIT $3`
+)
+
+type VectorStore struct {
+	engine             alloydbutil.PostgresEngine
+	embedder           embeddings.Embedder
+	tableName          string
+	schemaName         string
+	idColumn           string
+	metadataJsonColumn string
+	contentColumn      string
+	embeddingColumn    string
+	metadataColumns    []string
+	overwriteExisting  bool
+	k                  int
+	distanceStrategy   distanceStrategy
+}
+
+type BaseIndex struct {
+	name             string
+	indexType        string
+	distanceStrategy distanceStrategy
+	partialIndexes   []string
+}
+
+var _ vectorstores.VectorStore = &VectorStore{}
+
+// NewVectorStore creates a new VectorStore with options.
+func NewVectorStore(ctx context.Context, engine alloydbutil.PostgresEngine, embedder embeddings.Embedder, tableName string, opts ...AlloyDBVectoreStoresOption) (VectorStore, error) {
+	vs, err := applyAlloyDBVectorStoreOptions(engine, embedder, tableName, opts...)
+	if err != nil {
+		return VectorStore{}, err
+	}
+	return vs, nil
+}
+
+// AddDocuments adds documents to the Postgres collection, and returns the ids
+// of the added documents.
+func (vs *VectorStore) AddDocuments(ctx context.Context, docs []schema.Document, _ ...vectorstores.Option) ([]string, error) {
+	var texts []string
+	for _, doc := range docs {
+		texts = append(texts, doc.PageContent)
+	}
+	embeddings, err := vs.embedder.EmbedDocuments(ctx, texts)
+	if err != nil {
+		return nil, fmt.Errorf("failed embed documents: %w", err)
+	}
+	// If no ids provided, generate them.
+	ids := make([]string, len(texts))
+	for i, doc := range docs {
+		if val, ok := doc.Metadata["id"].(string); ok {
+			ids[i] = val
+		} else {
+			ids[i] = uuid.New().String()
+		}
+	}
+	// If no metadata provided, initialize with empty maps
+	metadatas := make([]map[string]interface{}, len(texts))
```

Comment:
nit: `any` is preferred over interface.
```suggestion
	metadatas := make([]map[string]any, len(texts))
```